packages/node/src/stellar/api.stellar.ts

```diff
@@ -72,61 +85,241 @@ export class StellarApi implements ApiWrapper<StellarBlockWrapper> {
   }
 
   getGenesisHash(): string {
-    return this.getChainId();
+    return this.chainId;
   }
 
   getSpecName(): string {
     return 'Stellar';
   }
 
-  async getEvents(height: number): Promise<SorobanRpc.GetEventsResponse> {
-    return this.client.getEvents({ startLedger: height, filters: [] });
-  }
+  private getTransactionApplicationOrder(eventId: string) {
+    // Right shift the ID by 12 bits to exclude the Operation Index
+    const shiftedId = BigInt(eventId.split('-')[0]) >> BigInt(12);
 
-  async fetchBlock(
-    blockNumber: number,
-    includeTx?: boolean,
-  ): Promise<StellarBlockWrapped> {
-    try {
-      const rawEvents = (await this.getEvents(blockNumber)).events;
+    // Create a mask for 20 bits to ignore the Ledger Sequence Number
+    const mask = BigInt((1 << 20) - 1);
 
-      const events = rawEvents.map((event) => ({
+    // Apply bitwise AND operation with the mask to get the Transaction Application Order
+    const transactionApplicationOrder = shiftedId & mask;
+    return Number(transactionApplicationOrder);
+  }
+
+  async getAndWrapEvents(height: number): Promise<SorobanEvent[]> {
+    const events = (
+      await this.sorobanClient.getEvents({ startLedger: height, filters: [] })
+    ).events;
+    return events.map((event) => {
+      const wrappedEvent = {
         ...event,
-        value: {
-          ...event.value,
-          get decoded() {
-            return scValToNative(xdr.ScVal.fromXDR(event.value.xdr, 'base64'));
+        ledger: null,
+        transaction: null,
+        operation: null,
+      } as SorobanEvent;
+
+      return wrappedEvent;
+    });
+  }
+
+  private async fetchEffectsForOperation(
+    operationId: string,
+  ): Promise<StellarEffect[]> {
+    const effects = (await this.api.effects().forOperation(operationId).call())
+      .records;
+    const wrappedEffects: StellarEffect[] = [];
+
+    effects.forEach((effect) => {
+      const wrappedEffect: StellarEffect = {
+        ...effect,
+        ledger: null,
+        transaction: null,
+        operation: null,
+      };
+
+      wrappedEffects.push(wrappedEffect);
+    });
+
+    return wrappedEffects;
+  }
+
+  private async fetchOperationsForTransaction(
+    transactionId: string,
+    applicationOrder: number,
+    sequence: number,
+  ): Promise<StellarOperation[]> {
+    const operations = (
+      await this.api.operations().forTransaction(transactionId).call()
+    ).records;
+
+    return Promise.all(
+      operations.map(async (op) => {
+        //There will be only one operation for soroban transaction
+        let events: SorobanEvent[] = [];
+        try {
+          events =
+            op.type.toString() === 'invoke_host_function'
+              ? (await this.getAndWrapEvents(sequence)).filter(
```

Comment:
We would make this same request multiple times in the same sequence if theres multiple operations in a sequence. We would be better to make the call once